# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: Gaurav Mahavir Banwat

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1899]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [1900]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [1901]:
#sets.info()
#sets.groupby('set_num')['num_parts'].head()
minifigs.head()

fig_num                             name  num_parts
0  fig-000001               Toy Store Employee          4
1  fig-000002                     Customer Kid          4
2  fig-000003            Assassin Droid, White          8
3  fig-000004                     Basic Figure          4
4  fig-000005  Captain America with Short Legs          3

In [1902]:
inventory_minifigs.head()

inventory_id     fig_num  quantity
0             3  fig-001549         1
1             4  fig-000764         1
2            19  fig-000555         1
3            25  fig-000574         1
4            26  fig-000842         1

In [1903]:
inventories.head()

id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [1904]:
def data_count(sets):
    return len(sets.index)


**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [1905]:
Q1=data_count(sets)
Q1

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [1906]:
Q2=int(np.round(sets['num_parts'].mean(),0))
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [1907]:
Q3=int(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [1908]:
Q4=sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [1909]:
Q5=sets.loc[sets['num_parts']==Q4,:]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [1910]:
def select_set_row(set_num):
    return sets.loc[sets['set_num']==set_num,:]

select_set_row('71040-1')

set_num           name  year  theme_id  num_parts
12144  71040-1  Disney Castle  2016       608       4081

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [1911]:
def select_set_numparts(number_of_parts):
    return sets.loc[sets['num_parts']==number_of_parts,:].sort_values(['year','set_num'])
#select_set_numparts('71040-1')

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [1912]:
select_set_numparts(22)

set_num                                               name  \
2899              282-1                     2 x 2 Sloping Roof Bricks, Red   
2901              282-2                    2 x 2 Sloping Roof Bricks, Blue   
18544           vwkit-1                                     Volkswagen Kit   
1365              152-1                                   Two Train Wagons   
7056              482-3  Slopes and Slopes Double 2 x 2, Red (The Build...   
...                 ...                                                ...   
12267           71361-6                                            Bob-omb   
738             10941-1                     Mickey & Minnie Birthday Train   
12296          71386-10                                        Bone Goomba   
17720  DOTS-MEGA-PACK-1           DOTS Collectible Tiles 41913-1 Mega Pack   
757             10960-1                                   Belle's Ballroom   

       year  theme_id  num_parts  
2899   1958       371         22  
2901   1958       371         22  
18544  1959       366         22  
1365   1966       235         22  
7056   1966       371         22  
...     ...       ...        ...  
12267  2020       690         22  
738    2021       641         22  
12296  2021       690         22  
17720  2021       688         22  
757    2022       504         22  

[114 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [1913]:
Q8A=sets['year'].min()
Q8A

1949

In [1914]:
Q8B=sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [1915]:
def select_set_year(year):
    return sets.loc[sets['year']==year,:].sort_values('set_num')
select_set_year(1949)

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

In [1916]:
themes.rename(columns={'name':'name_themes','id':'theme_id'},inplace=True)
sets_themes=sets.merge(themes)
#themes.head()
#sets.head()
sets_themes.head()

set_num                            name  year  theme_id  num_parts  \
0   001-1                           Gears  1965         1         43   
1   002-1  4.5V Samsonite Gears Motor Set  1965         1          3   
2  1030-1  TECHNIC I: Simple Machines Set  1985         1        211   
3  1038-1              ERBIE the Robo-Car  1985         1        120   
4  1039-1            Manual Control Set 1  1986         1         39   

  name_themes  parent_id  
0     Technic        NaN  
1     Technic        NaN  
2     Technic        NaN  
3     Technic        NaN  
4     Technic        NaN

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [1917]:
themes.rename(columns={'name':'name_themes','id':'theme_id'},inplace=True)
sets_themes=sets.merge(themes)
def theme_by_year(year):
    return sets_themes.loc[sets_themes['year']==year,['theme_id','name_themes']].sort_values('theme_id').rename(columns={'theme_id':'id'}).drop_duplicates ('id').reset_index(drop=True)


**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [1918]:
theme_by_year(1978)

id               name_themes
0     4            Expert Builder
1    67              Classic Town
2   189            Classic Castle
3   227                 Christmas
4   235                      4.5V
5   363                      Boat
6   364  Building Set with People
7   404                Hobby Sets
8   405                 Homemaker
9   412                   Airport
10  453                   Technic
11  469                 Basic Set
12  473              Supplemental
13  501                      Gear
14  505                 Basic Set
15  632                      Town

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [1919]:
sets_themes.rename(columns={'name':'name_sets'},inplace=True)
def theme_by_name(theme_name):
    return sets_themes.loc[sets_themes['name_themes']==theme_name,['set_num','name_sets','year','num_parts','theme_id','name_themes']].sort_values(by='set_num').reset_index(drop=True)
#sets_themes.head()

**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [1920]:
theme_by_name('Friends')

set_num                                          name_sets  year  \
0            10726-1                         Stephanie's Horse Carriage  2016   
1            10727-1                             Emma's Ice Cream Truck  2016   
2            10728-1                                   Mia's Vet Clinic  2016   
3            10746-1                                Mia's Farm Suitcase  2017   
4            10747-1               Andrea and Stephanie's Beach Holiday  2017   
..               ...                                                ...   ...   
477  FRNDSMAGJNGLE-1  LEGO Friends Magazine Issue 'Jungle adventures...  2014   
478  ICECREAMTRUCK-1                                    Ice Cream Truck  2020   
479    PUPPY-HOUSE-2                                        Puppy House  2019   
480     TRUFRIENDS-1                                       Wishing Well  2016   
481    WISHINGWELL-1                                       Wishing Well  2013   

     num_parts  theme_id name_themes  
0           58       595     Friends  
1          136       595     Friends  
2          182       595     Friends  
3          106       595     Friends  
4          144       595     Friends  
..         ...       ...         ...  
477         28       494     Friends  
478         60       494     Friends  
479         72       494     Friends  
480         33       494     Friends  
481         28       494     Friends  

[482 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [1921]:
def theme_by_setnum(set_num):
    ind=sets_themes.loc[sets_themes['set_num']==set_num].index.values.astype(int)[0]
    return sets_themes.loc[ind,'name_themes']

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [1922]:
theme_by_setnum("75810-1")

'Stranger Things'

In [1923]:
minifigs.head()

fig_num                             name  num_parts
0  fig-000001               Toy Store Employee          4
1  fig-000002                     Customer Kid          4
2  fig-000003            Assassin Droid, White          8
3  fig-000004                     Basic Figure          4
4  fig-000005  Captain America with Short Legs          3

**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [1924]:
sets_c=sets.copy()
inventories_c=inventories.copy()
inventory_minifigs_c=inventory_minifigs.copy()
minifigs_c=minifigs.copy()



sets_c=sets_c.rename(columns={'name':'set_name','num_parts':'set_num_parts'})
inventories_c=inventories.rename(columns={'id':'inventory_id','version':'inventory_version'})
inventory_minifigs_c=inventory_minifigs.rename(columns={'quantity':'inventory_minifigs_quantity'})
minifigs_c=minifigs_c.rename(columns={'name':'fig_name','num_parts':'fig_num_parts'})

df1=sets_c.merge(inventories_c)
df2=df1.merge(inventory_minifigs_c)
df3=df2.merge(minifigs_c)
df3.head()

set_num             set_name  year  theme_id  set_num_parts  inventory_id  \
0  0011-2    Town Mini-Figures  1979        67             12          5087   
1  1064-1      Dacta Buildings  1981       533            419          8812   
2  1067-1   Community Vehicles  1985       507            257           859   
3  1490-1            Town Bank  1988        67            196          6787   
4   381-2  Police Headquarters  1979        67            372         42209   

   inventory_version     fig_num  inventory_minifigs_quantity  \
0                  1  fig-000225                            1   
1                  1  fig-000225                            1   
2                  1  fig-000225                            1   
3                  1  fig-000225                            1   
4                  2  fig-000225                            1   

                                            fig_name  fig_num_parts  
0  Policeman, Black Suit with Pocket and Badge, W...              4  
1  Policeman, Black Suit with Pocket and Badge, W...              4  
2  Policeman, Black Suit with Pocket and Badge, W...              4  
3  Policeman, Black Suit with Pocket and Badge, W...              4  
4  Policeman, Black Suit with Pocket and Badge, W...              4

In [1925]:
MFEP=df3.loc[df3['set_name']=='Millennium Falcon',:]
#MFEP.head()
#MFEP['set_num_parts'].max()
l_MFEP_df=MFEP.loc[MFEP['set_num_parts']==MFEP['set_num_parts'].max(),:]
l_MFEP_df.head()

set_num           set_name  year  theme_id  set_num_parts  \
13447  75192-1  Millennium Falcon  2017       171           7541   
13658  75192-1  Millennium Falcon  2017       171           7541   
13683  75192-1  Millennium Falcon  2017       171           7541   
13713  75192-1  Millennium Falcon  2017       171           7541   
13742  75192-1  Millennium Falcon  2017       171           7541   

       inventory_id  inventory_version     fig_num  \
13447         19670                  1  fig-001714   
13658         19670                  1  fig-002057   
13683         19670                  1  fig-002544   
13713         19670                  1  fig-001810   
13742         19670                  1  fig-002514   

       inventory_minifigs_quantity  \
13447                            1   
13658                            1   
13683                            1   
13713                            1   
13742                            1   

                                              fig_name  fig_num_parts  
13447                            Chewbacca, Dark Brown              3  
13658                               Rey, Dark Tan Robe              4  
13683                                             BB-8              2  
13713                           Finn, Black Undershirt              4  
13742  C-3PO, Pearl Gold, Colorful Wires, Printed Legs              3

In [1927]:
l_MFEP_df['set_num']

13447    75192-1
13658    75192-1
13683    75192-1
13713    75192-1
13742    75192-1
13914    75192-1
13967    75192-1
13968    75192-1
Name: set_num, dtype: object

In [1928]:
l_MFEP_df['inventory_version']

13447    1
13658    1
13683    1
13713    1
13742    1
13914    1
13967    1
13968    1
Name: inventory_version, dtype: int64

**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [1929]:
Q13=l_MFEP_df['inventory_minifigs_quantity'].sum()
Q13

8

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [1930]:
#def get_minifigs(set_number,version_number):
#    tdf=largest_MFEP_df.loc[largest_MFEP_df['set_num']==set_number,:]
#    return tdf.loc[tdf['inventory_version'==version_number],['set_num','fig_num','fig_name','fig_num_parts']].sort_values('fig_num').reset_index(drop=True)
    

    
def get_minifigs(set_num,version_num):
    tdf=df3.loc[df3['set_num']==set_num,:]
    tdf2=tdf.loc[tdf['inventory_version']==version_num,:]
    tdf3=tdf2.loc[:,['set_num','fig_num','fig_name','fig_num_parts']]
    tdf4=tdf3.sort_values(by='fig_num')
    tdf5=tdf4.reset_index(drop=True)
    return tdf5
    #return tdf.loc[tdf['inventory_version'==version_num],['set_num','fig_num','fig_name','fig_num_parts']].sort_values('fig_num').reset_index(drop=True)
    

In [1931]:
get_minifigs('10217-1', 2)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [1932]:
#from Millennium Falcon Prep Exercise:
#set_num='75192-1' and inventory_version =1

Q15= get_minifigs('75192-1', 1)
Q15

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [1933]:
get_minifigs('10217-1',1)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

In [1934]:
get_minifigs('10217-1',2)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [1935]:
def minifigs_from_themes(theme_id):
    sdf1=df3.loc[df3['theme_id']==theme_id,['fig_num','fig_name']]
    sdf2=sdf1.sort_values(by='fig_num')
    sdf3=sdf2.drop_duplicates()
    sdf4=sdf3.reset_index(drop=True)
    return sdf4

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [1936]:
#from Millennium Falcon Prep Exercise 'theme_id' =171

#df3.head()
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [1937]:
def sets_from_minifig(minifig_number):
    rdf1=df3.loc[df3['fig_num']==minifig_number,['set_num', 'inventory_version','set_name', 'fig_num', 'fig_name']]
    rdf2=rdf1.rename(columns={'inventory_version':'version', 'set_name':'name_sets'})
    rdf3=rdf2.reset_index(drop=True)
    return rdf3

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [1938]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [1939]:
#Merging the Dataframes and renaming the columns as required
#sets_c=sets.copy()
#inventories_c=inventories.copy()
#inventory_minifigs_c=inventory_minifigs.copy()
#minifigs_c=minifigs.copy()




DF1=themes.merge(sets_c)
DF2=DF1.merge(inventories_c)
DF3=DF2.merge(inventory_parts)
parts.rename(columns={'name':'part_name'},inplace=True)
DF4=DF3.merge(parts)
colors.rename(columns={'id':'color_id','name':'color_name'},inplace=True)
DF5=DF4.merge(colors)
DF5.head()

theme_id name_themes  parent_id set_num  \
0         1     Technic        NaN   001-1   
1       235        4.5V      233.0   100-1   
2       235        4.5V      233.0   100-2   
3       301       Other        NaN   371-2   
4       365      System        NaN   310-1   

                                 set_name  year  set_num_parts  inventory_id  \
0                                   Gears  1965             43         24696   
1  4.5V Motor with Wheels (Small Version)  1966             16         15097   
2  4.5V Motor with Wheels (Large Version)  1966             27          9538   
3                     Motorized Truck Set  1967            314          5787   
4                     Motorized Truck Set  1969            185         25382   

   inventory_version part_num  color_id  quantity is_spare  \
0                  1     132a         7         4        f   
1                  1     132a         7         2        f   
2                  1     132a         7         4        f   
3                  1     132a         7        14        f   
4                  1     132a         7        16        f   

                       part_name  part_cat_id part_material  color_name  \
0  Tyre Smooth Old Style - Small           29        Rubber  Light Gray   
1  Tyre Smooth Old Style - Small           29        Rubber  Light Gray   
2  Tyre Smooth Old Style - Small           29        Rubber  Light Gray   
3  Tyre Smooth Old Style - Small           29        Rubber  Light Gray   
4  Tyre Smooth Old Style - Small           29        Rubber  Light Gray   

      rgb is_trans  
0  9BA19D        f  
1  9BA19D        f  
2  9BA19D        f  
3  9BA19D        f  
4  9BA19D        f

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [1940]:
def set_parts(set_num,i_version):
    qdf1=DF5.loc[DF5['set_num']==set_num,:]
    qdf2=qdf1.loc[qdf1['inventory_version']==i_version,['set_num', 'inventory_version', 'set_name', 'year', 'name_themes', 'part_num', 'part_name', 'color_name','quantity','is_spare']]
    qdf3=qdf2.rename(columns={'inventory_version':'version','set_name':'name_sets','color_name':'name_color','part_name':'name_part'})
    qdf4=qdf3.sort_values(by=['part_num','name_color'])
    qdf5=qdf4.reset_index(drop=True)
    return qdf5
    

In [1941]:
# As per Millennium Falcon Prep Exercise:
#set_num= '75192-1'
#i_version= 1
set_parts('75192-1',1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
725  Light Bluish Gray        70        f  
726              Black         6        f  
727   Dark Bluish Gray        25        f  
728  Light Bluish Gray         2        f  
729  Light Bluish Gray         3        f  

[730 rows x 10 columns]

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets? 
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list. 
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
